src/core/dbus-execute.c

```diff
@@ -2087,52 +2124,53 @@ int bus_exec_context_set_transient_property(
                 return 1;
 
         } else if (STR_IN_SET(name, "RuntimeDirectory", "StateDirectory", "CacheDirectory", "LogsDirectory", "ConfigurationDirectory")) {
+                ExecDirectoryType i = exec_directory_type_from_string(name);
                 _cleanup_strv_free_ char **l = NULL;
                 char **p;
 
+                assert(i >= 0 && i < _EXEC_DIRECTORY_TYPE_MAX);
+
                 r = sd_bus_message_read_strv(message, &l);
                 if (r < 0)
                         return r;
 
                 STRV_FOREACH(p, l) {
-                        if (!path_is_normalized(*p))
-                                return sd_bus_error_setf(error, SD_BUS_ERROR_INVALID_ARGS, "%s= path is not normalized: %s", name, *p);
+                        char *q = *p + (**p == '-');
+
+                        if (!path_is_normalized(q))
+                                return sd_bus_error_setf(error, SD_BUS_ERROR_INVALID_ARGS, "%s= path is not normalized: %s", name, q);
 
-                        if (path_is_absolute(*p))
-                                return sd_bus_error_setf(error, SD_BUS_ERROR_INVALID_ARGS, "%s= path is absolute: %s", name, *p);
+                        if (path_is_absolute(q))
+                                return sd_bus_error_setf(error, SD_BUS_ERROR_INVALID_ARGS, "%s= path is absolute: %s", name, q);
 
-                        if (path_startswith(*p, "private"))
-                                return sd_bus_error_setf(error, SD_BUS_ERROR_INVALID_ARGS, "%s= path can't be 'private': %s", name, *p);
+                        if (path_startswith(q, "private"))
+                                return sd_bus_error_setf(error, SD_BUS_ERROR_INVALID_ARGS, "%s= path can't be 'private': %s", name, q);
                 }
 
                 if (!UNIT_WRITE_FLAGS_NOOP(flags)) {
-                        char ***dirs = NULL;
-                        ExecDirectoryType i;
-
-                        for (i = 0; i < _EXEC_DIRECTORY_TYPE_MAX; i++)
-                                if (streq(name, exec_directory_type_to_string(i))) {
-                                        dirs = &c->directories[i].paths;
-                                        break;
-                                }
-
-                        assert(dirs);
+                        ExecDirectories *ed = &c->directories[i];
 
                         if (strv_isempty(l)) {
-                                *dirs = strv_free(*dirs);
+                                exec_directories_clear(ed);
                                 unit_write_settingf(u, flags, name, "%s=", name);
-                        } else {
-                                _cleanup_free_ char *joined = NULL;
+                        } else
+                                STRV_FOREACH(p, l) {
+                                        bool ignore = false;
+                                        char *q = *p;
```

Comment:
`const` here too?

Reply:
It is possible, but `.path` is `char *`. Casting from `char*` to `const char*` here and revert it at the later substitution may be not good, I think.